src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/DecryptorPalOpenSsl.Decrypt.cs

```diff
@@ -0,0 +1,124 @@
+    {
+        public sealed override ContentInfo TryDecrypt(
+            RecipientInfo recipientInfo,
+            X509Certificate2 cert,
+            X509Certificate2Collection originatorCerts,
+            X509Certificate2Collection extraStore,
+            out Exception exception)
+        {
+            Debug.Assert(recipientInfo != null);
+            Debug.Assert(cert != null);
+            Debug.Assert(originatorCerts != null);
+            Debug.Assert(extraStore != null);
+
+            string recipiendKekAlgo = recipientInfo.KeyEncryptionAlgorithm.Oid.Value;
+            string certOid = cert.GetKeyAlgorithm();
+
+            ContentInfo decryptedContent = null;
+            using (SafeBioHandle ecm = null)
+            {
+                RecipientInfoType type = recipientInfo.Type;
+                switch (type)
+                {
+                    case RecipientInfoType.KeyTransport:
+                        decryptedContent = TryDecryptTrans(cert, out exception);
+                        break;
+                        
+                    case RecipientInfoType.KeyAgreement:
+                        // OpenSSL doesn't seem to support KeyAgreement directly, the only way to work around this seems to be to extract
+                        // the information by parsing it and to store the appropriate keys to call the EVP_PKEY_derive method on them
+                        // to obtain the secret and proceed with manual decryption. For further documentation see:
+                        // https://wiki.openssl.org/index.php/EVP_Key_Agreement
+                        exception = new PlatformNotSupportedException();
+                        break;
+
+                    default:
+                        // As we commented in the Windows implementation:
+                        // Since only the framework can construct RecipientInfo's, we're at fault if we get here. So it's okay to assert and throw rather than
+                        // returning to the caller.
+                        Debug.Fail($"Unexpected RecipientInfoType: {type}");
+                        throw new NotSupportedException();
+                }
+            }
+
+            if (exception != null)
+                return null;
+
+            return decryptedContent;
+        }
+
+        private ContentInfo TryDecryptTrans(X509Certificate2 recipientCert, out Exception exception)
+        {
+            // As of now, KeyTransport recipients have RSA as their only choice for the symmetric key encryption,
+            // but to future proof, might as well check so that we don't try to get a private RSA key from something that doesn't have one.
+            if (recipientCert.GetKeyAlgorithm() != Oids.Rsa)
+            {
+                exception = new PlatformNotSupportedException();
+                return null;
+            }
+
+            RSAOpenSsl kekAlgoPKey = recipientCert.GetRSAPrivateKey() as RSAOpenSsl;
```

Comment:
A couple of points here:
* The call to GetRSAPrivateKey should be in a using
* You could remove the previous GetKeyAlgorithm() call and just use whether or not GetRSAPrivateKey returned null
* In the future, if GetRSAPrivateKey returned something other than an RSAOpenSsl you'd end up with a NullReferenceException later.  You may as well hard-cast and get the casting exception.  (Or check the output of the as for null, and use *that* to set the PNSE)